### PR TITLE
DAOS-7492 dfuse: stop progress thread on error

### PR DESCRIPTION
**Pull request by johannlombardi**
Properly stop dfuse progress thread on errors to avoid DER_NO_PERM
error from daos_eq_lib_fini when hitting errors during ll_loop_fn.

Signed-off-by: Johann Lombardi <johann.lombardi@intel.com>

### CODE REVIEW

File: src/client/dfuse/dfuse_core.c [hunk 1]
```diff
@@ -979,13 +979,14 @@ dfuse_start(struct dfuse_projection_info *fs_handle,
...
+	rc = dfuse_launch_fuse(fs_handle, fuse_ops, &args);
+	D_FREE(fuse_ops);
+	if (!rc) {
+		(void)dfuse_fs_fini(fs_handle);
 		DFUSE_TRA_ERROR(fs_handle, "Unable to register FUSE fs");
-		D_GOTO(err_ie_remove, rc = -DER_INVAL);
+		return -DER_INVAL;
 	}
 
-	D_FREE(fuse_ops);
```
Comment (ashleypittman) [inline src/client/dfuse/dfuse_core.c:987]: Won't this introduce a memory leak?

**Comment by ashleypittman** [comment]
I'm assuming this works because it avoids the decref(), so dfuse will not try and cleanup and simply leak rather than fail when it comes to shutting down daos/cart?  I'm fine with this if it helps in this case although I intend to try dfuse with the fault injection testing soon which will lead to a number of cleanups in this area.

File: src/client/dfuse/dfuse_core.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/dfuse/dfuse_core.c:987]: the free was moved to lone 986

**Comment by johannlombardi** [comment]
The decref will be done as part of dfuse_fs_fini(), no? What i see on my side is that the progress thread is not stopped in the case of early failure in the fuse loop.

**Comment by ashleypittman** [comment]
> The decref will be done as part of dfuse_fs_fini(), no? What i see on my side is that the progress thread is not stopped in the case of early failure in the fuse loop.

Yes, it should be.  dfuse_fs_fini() is a belt-and-braces function to proactively destroy any leaked references or descriptions on shutdown, we use it to detect discrepancies between kernel references and what dfuse thinks it keeps, so whilst dfuse_fs_fini() will free it I would consider it an error if we were using that function in non-error cases.  For this though I think it's fine.

**Comment by ashleypittman** [comment]
https://github.com/daos-stack/daos/pull/5863 changes the -DER_NO_PERM to -DER_BUSY and improves the error messages, although does nothing to avoid the error.

**Review by ashleypittman — APPROVED** [review]

**Comment by ashleypittman** [comment]
I've incorporated this into https://github.com/daos-stack/daos/pull/5866

**Pull request merged**